FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/transport/src/tls_freertos.c

```diff
@@ -0,0 +1,604 @@
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+TlsTransportStatus_t TLS_FreeRTOS_Connect( NetworkContext_t * pNetworkContext,
+                                           const char * pHostName,
+                                           uint16_t port,
+                                           const NetworkCredentials_t * pNetworkCredentials,
+                                           uint32_t receiveTimeoutMs,
+                                           uint32_t sendTimeoutMs )
+{
+    TlsTransportStatus_t returnStatus = TLS_TRANSPORT_SUCCESS;
+    BaseType_t socketStatus = 0;
+
+    if( ( pNetworkContext == NULL ) ||
+        ( pHostName == NULL ) ||
+        ( pNetworkCredentials == NULL ) )
+    {
+        LogError( ( "Arguments cannot be NULL. pNetworkContext=%p, "
+                    "pHostName=%p, pNetworkCredentials=%p.",
+                    pNetworkContext,
+                    pHostName,
+                    pNetworkCredentials ) );
+        returnStatus = TLS_TRANSPORT_INVALID_PARAMETER;
+    }
+    else if( ( pNetworkCredentials->pRootCa == NULL ) )
+    {
+        LogError( ( "pRootCa cannot be NULL." ) );
+        returnStatus = TLS_TRANSPORT_INVALID_PARAMETER;
+    }
+
+    /* Create a FreeRTOS+TCP socket. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        socketStatus = Sockets_Connect( &( pNetworkContext->tcpSocket ),
+                                        pHostName,
+                                        port,
+                                        receiveTimeoutMs,
+                                        sendTimeoutMs );
+
+        if( socketStatus != 0 )
+        {
+            LogError( ( "Failed to connect to %s with error %d.",
+                        pHostName,
+                        socketStatus ) );
+            returnStatus = TLS_TRANSPORT_CONNECT_FAILURE;
+        }
+    }
+
+    /* Initialize mbedtls. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        returnStatus = initMbedtls();
+    }
+
+    /* Perform TLS handshake. */
+    if( returnStatus == TLS_TRANSPORT_SUCCESS )
+    {
+        returnStatus = tlsSetup( pNetworkContext, pHostName, pNetworkCredentials );
+    }
+
+    /* Clean up on failure. */
+    if( returnStatus != TLS_TRANSPORT_SUCCESS )
+    {
+        if( pNetworkContext->tcpSocket != FREERTOS_INVALID_SOCKET )
+        {
+            FreeRTOS_closesocket( pNetworkContext->tcpSocket );
+        }
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) Connection to %s established.",
+                   pNetworkContext,
+                   pHostName ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+void TLS_FreeRTOS_Disconnect( NetworkContext_t * pNetworkContext )
+{
+    BaseType_t socketStatus = 0;
+
+    socketStatus = ( BaseType_t ) mbedtls_ssl_close_notify( &( pNetworkContext->sslContext.context ) );
```

Comment:
(Nit) A comment here stating that TLS session is being terminated would be good to have

Reply:
Ok Sure